### PR TITLE
Bed Fix

### LINKED ISSUE #2656: Sleeping Citizen - Sharing Beds
When the citizen is sleeping in a bed.  You the player is able to sleep in the same bed.   MC should prevent you from sleep in the bed when a citizen is in the bed.

### PR DESCRIPTION
**Pull request by ghost**
Fixed Sleeping in same Bed as Citizen Bug

Closes #2656 

# Changes proposed in this pull request:
-Updates EventHandler.java to check OnActivated Event for Player to see if player is trying to sleep in bed with a citizen in it, and cancels the event if true.

Review please

### CODE REVIEW

**Review by LDTTeam-Buildsystem — DISMISSED** [review]
#### Analysis Complete
##### Statistics:
|        | Count | New | Old | Diff |
| ------ |:-----:|:---:|:---:|:----:|
| Total  |   0   |  0  |  0  |  0   |
| Errors |   0   |  0  |  0  |  0   |
##### More Information:

**Review by LDTTeam-Buildsystem — DISMISSED** [review]
#### Analysis Complete
##### Statistics:
|        | Count | New | Old | Diff |
| ------ |:-----:|:---:|:---:|:----:|
| Total  |   0   |  0  |  0  |  0   |
| Errors |   0   |  0  |  0  |  0   |
##### More Information:

File: src/main/java/com/minecolonies/coremod/event/EventHandler.java [hunk 1]
```diff
@@ -309,6 +314,22 @@ else if ("pmardle".equalsIgnoreCase(event.getEntityPlayer().getName())
...
             }
+             //Checks to see if player tries to sleep in a bed belonging to a Citizen, ancels the event, and Notifies Player that bed is occuppied
+            if(world.getBlockState(event.getPos()).getBlock().isBed(world.getBlockState(event.getPos()), world, event.getPos(), player))
+            {
+                BlockPos bd = event.getPos();
+                if(world.getBlockState(event.getPos()).getBlock().isBedFoot(world, event.getPos())) {
+                    bd = bd.offset(world.getBlockState(event.getPos()).getValue(BlockBed.FACING));
+                }
+                //Searches through the nearest Colonie's Citizen and sees if the bed belongs to a Citizen, and if the Citizen is asleep
+                final ArrayList<CitizenData> cits = new ArrayList<>(ColonyManager.getClosestColony(world, event.getPos()).getCitizenManager().getCitizens());
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/event/EventHandler.java]: List<CitizenData> cits = new ArrayList<>();
Also, better save the colony and check for null first, if colony is null no need for all of this.

File: src/main/java/com/minecolonies/coremod/event/EventHandler.java [hunk 2]
```diff
@@ -309,6 +314,22 @@ else if ("pmardle".equalsIgnoreCase(event.getEntityPlayer().getName())
...
+                for(int i = 0; i < cits.size(); i++){
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/event/EventHandler.java]: for each?

**Review by Raycoms — DISMISSED** [review]
Nice changes, a few small additions:

Can you please update the formatting to fit the general style?
Especially the bracket style, there is a formatter xml file in the repo you can download and import with intellij

**Comment by ghost** [comment]
Will Do. Give me a day or so to fix things over.

File: src/main/java/com/minecolonies/coremod/event/EventHandler.java [hunk 3]
```diff
@@ -309,6 +314,22 @@ else if ("pmardle".equalsIgnoreCase(event.getEntityPlayer().getName())
             {
                 return;
             }
+             //Checks to see if player tries to sleep in a bed belonging to a Citizen, ancels the event, and Notifies Player that bed is occuppied
+            if(world.getBlockState(event.getPos()).getBlock().isBed(world.getBlockState(event.getPos()), world, event.getPos(), player))
+            {
+                BlockPos bd = event.getPos();
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/event/EventHandler.java]: Move it before the `if` if possible (and use it whenever possible), also rename to something more understandable

Comment (Nightenom) [reply src/main/java/com/minecolonies/coremod/event/EventHandler.java]: edit: move it on 289 where are finals for this event, on 322 create a new variable with an appropriate name

**Review by LDTTeam-Buildsystem — DISMISSED** [review]
#### Analysis Complete
##### Statistics:
|        | Count | New | Old | Diff |
| ------ |:-----:|:---:|:---:|:----:|
| Total  |   0   |  0  |  0  |  0   |
| Errors |   0   |  0  |  0  |  0   |
##### More Information:

**Review by LDTTeam-Buildsystem — DISMISSED** [review]
#### Analysis Complete
##### Statistics:
|        | Count | New | Old | Diff |
| ------ |:-----:|:---:|:---:|:----:|
| Total  |   0   |  0  |  0  |  0   |
| Errors |   0   |  0  |  0  |  0   |
##### More Information:

**Review by LDTTeam-Buildsystem — APPROVED** [review]
#### Analysis Complete
##### Statistics:
|        | Count | New | Old | Diff |
| ------ |:-----:|:---:|:---:|:----:|
| Total  |   0   |  0  |  0  |  0   |
| Errors |   0   |  0  |  0  |  0   |
##### More Information:

**Review by LDTTeam-Buildsystem — DISMISSED** [review]
#### Analysis Complete
##### Statistics:
|        | Count | New | Old | Diff |
| ------ |:-----:|:---:|:---:|:----:|
| Total  |   0   |  0  |  0  |  0   |
| Errors |   0   |  0  |  0  |  0   |
##### More Information:

File: src/main/java/com/minecolonies/coremod/event/EventHandler.java [hunk 4]
```diff
@@ -286,6 +293,9 @@ public void onPlayerInteract(@NotNull final PlayerInteractEvent.RightClickBlock
     {
         final EntityPlayer player = event.getEntityPlayer();
         final World world = event.getWorld();
+        BlockPos bedBlockPos = event.getPos();
+        final List<CitizenData> citizenList;
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/event/EventHandler.java]: I would get the event pos here yes, but I wouldn't put the world and citizen list here. Else we query the colony for each right click of a player, this is terrible performance wise

File: src/main/java/com/minecolonies/coremod/event/EventHandler.java [hunk 5]
```diff
@@ -315,6 +325,27 @@ else if ("pmardle".equalsIgnoreCase(event.getEntityPlayer().getName())
                 return;
             }
 
+                //Checks to see if player tries to sleep in a bed belonging to a Citizen, ancels the event, and Notifies Player that bed is occuppied
+                if (colony != null && world.getBlockState(event.getPos()).getBlock().isBed(world.getBlockState(event.getPos()), world, event.getPos(), player))
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/event/EventHandler.java]: So first check for  bed and then check in the if for colony.

File: src/main/java/com/minecolonies/coremod/event/EventHandler.java [hunk 6]
```diff
@@ -315,6 +325,27 @@ else if ("pmardle".equalsIgnoreCase(event.getEntityPlayer().getName())
...
+                {
+                   citizenList = ColonyManager.getClosestColony(world, event.getPos()).getCitizenManager().getCitizens();
+                    if (world.getBlockState(event.getPos()).getBlock().isBedFoot(world, event.getPos()))
+                    {
+                     bedBlockPos = bedBlockPos.offset(world.getBlockState(event.getPos()).getValue(BlockBed.FACING));
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/event/EventHandler.java]: this need indent

File: src/main/java/com/minecolonies/coremod/event/EventHandler.java (hunk 4 above)
Comment (ghost) [reply src/main/java/com/minecolonies/coremod/event/EventHandler.java]: The World object was already implemented in the code that way at the start, as it's used by several other events when the Player Right-Clicks. I'll move the Citizenlist object thought.

Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/event/EventHandler.java]: sorry, I meant colony not world xD

**Comment by ghost** [comment]
Looks like it's all here. Codacy Likes it, format is good, Just waiting on approval.

File: src/main/java/com/minecolonies/coremod/event/EventHandler.java [hunk 7]
```diff
@@ -314,6 +322,30 @@ else if ("pmardle".equalsIgnoreCase(event.getEntityPlayer().getName())
...
                 return;
             }
+            if (world.getBlockState(event.getPos()).getBlock().isBed(world.getBlockState(event.getPos()), world, event.getPos(), player))
+            {
+
+                final IColony colony = ColonyManager.getIColony(world, event.getPos());
+                //Checks to see if player tries to sleep in a bed belonging to a Citizen, ancels the event, and Notifies Player that bed is occuppied
+                if (colony != null)
+                {
+                    final List<CitizenData> citizenList = ColonyManager.getClosestColony(world, event.getPos()).getCitizenManager().getCitizens();
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/event/EventHandler.java]: why don't you get a real colony above already, there is a ColonyManager.getColonyByPosFromWorld() then you don't have to get the colony twice.

Comment (ghost) [reply src/main/java/com/minecolonies/coremod/event/EventHandler.java]: I don't see that function anywhere. I have GetClosestColony, and GetClosestIColony

**Comment by Raycoms** [comment]
:shipit:

[Approved with PullApprove](https://pullapprove.com/Minecolonies/minecolonies/pull-request/2936/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Minecolonies/minecolonies)

**Pull request merged**